Sprint 12

Peppa-Akk opened the pull request:
- Add Network Layer for load images from Unsplash API;
- Add Network Layer for Like Button;
- Open large image when tap on cell with image;
- Added the possibility of leaving the account.

Peppa-Akk committed:
-Create Navigation Controller and Tab Controller;
- Add action with touch on cell with image;
- Add new assets.

Peppa-Akk committed:
Sprint 9:
- Add TabBarController;
- Add View with sizeble image;
- Add share button for image;
- Recode UI on Profile View.

Peppa-Akk committed:
Fix font

Peppa-Akk committed:
Some fix from pull request

Peppa-Akk committed:
Sprint 10:
- Add webView;
- Add request to unsplash authorize;
- Add fetch code in authorization request.

Peppa-Akk committed:
Sprint 10:
- Add webView;
- Add authorize request;
- Add fetch code in authorize request.

Peppa-Akk committed:
Add some features from 10-th sprint:
- Add Web View Controller;
- Add UIProgressView with KVO.
- Authorization with OAuth2;
- Save Bearer token.

Peppa-Akk committed:
Add some features from 10-th sprint:
- Add Web View Controller;
- Add UIProgressView with KVO;
- Authorization with OAuth2;
- Save Bearer token.

Peppa-Akk committed:
Add SplashViewController:
- Check UserDefaults on Bearer token
(if token saved in UD prepare to profile flow)
(if token unsaved in UD prepare to authorization flow)

Peppa-Akk committed:
Sprint 11:
- Add ProgressHUD to project with SPM;
- fix race condition.

Peppa-Akk committed:
Sprint 11:
- Add networl layer with ProfileService (Fetch next data: first name, last name, username and bio);
- Update data in ProfileViewController.

Peppa-Akk committed:
Sprint 11:
- Replaced UI with Keychain;
- Update new API KVO;
- Add ProfileImageService;
- Set avatar photo with Kingfisher.

Peppa-Akk committed:
Sprint 11:
- Key window started from code now;
- UI of SplashViewController and TabBarController in code;
- Update ProgressHUD from 13.6.0 to 13.6.1.

Peppa-Akk committed:
Sprint 12:
- Add Photo Model (without CodingKeys);
- Start ImageListService.

Peppa-Akk committed:
Sprint 12:
- End Network layer ImageListService;
- Update method tableView(... willDisplay cell: UITableViewCell ...).

Peppa-Akk committed:
Sprint 12:
- Add Network Layer for load images from Unsplash API;
- Add Network Layer for Like Button;
- Open large image when tap on cell with image;
- Added the possibility of leaving the account

helecta commented on ImageFeed/Auth/WebViewController.swift:
```diff
@@ -0,0 +1,97 @@
+    
+    // MARK: - Funcs
+    private func updateProgress() {
+        progressView.progress = Float(webView.estimatedProgress)
```
если использовать progressView.setProgress(…), то можно сделать заполнение прогресс бара более плавным.

helecta replied:
стоит установить начальное значение 0, это лучше всего сделать во viewDidLoad

Peppa-Akk replied:
Изменено в 13-м спринте

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -7,68 +7,160 @@
+    private var imageServiceObserver: NSObjectProtocol?
+    private var alertPresenter: AlertPresenter?
+    
+    private let ShowSingleImageSegueIdentifier = "ShowSingleImage"
```
Такие константы стоит называть со строчной буквы.  Да, в учебнике было так, но все-таки правильно называть со строчной. С заглавной можно называть только типы и протоколы.
Почитать подробнее можно  [тут](https://github.com/RedMadRobot/RMR-swift-style-guide#naming)  или  [тут](https://github.com/tutu-ru/swift-style-guide#%D0%B8%D0%BC%D0%B5%D0%BD%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5)

helecta commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -7,68 +7,160 @@
+            cell.cellImage.kf.setImage(with: url, placeholder: placeholderImage)
+        }
+        //set date
+        let textDate = dateFormatter.string(from: photo.createdAt ?? Date())
```
Тут не стоит использовать по умолчанию текущую дату, т.к это будет уже отображение неправильных данных, по заданию тут должен быть пустой лейбл:
> либо в структуре Photo поле `createdAt` опционально, в ячейке лейбл пустой

helecta commented on ImageFeed/Models/PhotosModel.swift:
```diff
@@ -0,0 +1,44 @@
+    let urls: UrlsResult
+    let likedByUser: Bool
+    
+    enum CodingKeys: String, CodingKey {
```
Еще можно таким образом настроить JSONDecoder: https://nemecek.be/blog/90/how-to-decode-snake-case-with-codable  
Тогда CodingKeys будет не нужно

Peppa-Akk replied:
Если честно, то такой метод больше нравится =)

helecta commented on ImageFeed/Profile/ProfileViewController.swift:
```diff
@@ -0,0 +1,178 @@
+    }
+    
+    private func updateProfileDetails(profile: Profile) {
+        self.nameLabel.text = profile.name
```
Тут `self` необязательно использовать, поэтому лучше его убрать.
Подробнее можешь почитать в [код стайле](https://github.com/tutu-ru/swift-style-guide#комментарии)

Peppa-Akk replied:
Изменено в 13-м спринте

helecta commented on ImageFeed/Services/ImageListService.swift:
```diff
@@ -0,0 +1,141 @@
+        
+        currentTask?.cancel()
+        
+        let nextPage = lastLoadedPage == nil ? 1 : lastLoadedPage! + 1
```
Тут можно переписать без force unwrap.
Force операции могут приводить к падению приложения, считается что падение приложения хуже, чем не рабочее состояние. Да, тут падение невозможно, но стоит взять себе за привычку если можно без force unwrap написать, то нужно писать без.

Как вариант можно сделать дефолтное значение lastLoadedPage:
`var lastLoadedPage: Int = 1`

helecta commented on ImageFeed/Services/ImageListService.swift:
```diff
@@ -0,0 +1,141 @@
+            id: photoResult.id,
+            size: CGSize(width: Double(photoResult.width), height: Double(photoResult.height)),
+            createdAt: ISO8601DateFormatter().date(
+                from: photoResult.createdAt ?? DateFormatter().string(from: Date())
```
По заданию форматтеры должны создаваться один раз, тут у тебя на каждую фото будет создан форматтер:

> Все форматтеры созданы один раз для использования. 
>		[Статья про стоимость date formatter](https://sarunw.com/posts/how-expensive-is-dateformatter/)

helecta replied:
аналогично для ISO8601DateFormatter()

Peppa-Akk replied:
Т.е. достаточно объявить в классе инициализацию форматтера?

helecta commented on ImageFeed/Services/OAuth2Service.swift:
```diff
@@ -0,0 +1,89 @@
+import Foundation
+
+fileprivate let DefaultBaseURL = URL(string: "https://api.unsplash.com")!
```
Такие константы стоит называть со строчной буквы.

helecta commented on ImageFeed/Services/ImageListService.swift:
```diff
@@ -0,0 +1,141 @@
+                if let index = self.photos.firstIndex(where: { $0.id == photoId }) {
+                    let photo = self.photos[index]
+                    
+                    let newPhoto = Photo(
```
лучше создавать newPhoto из полученного результата:
```
guard let photo = photoResult.photo,
                let index = self.photos.firstIndex(where: { $0.id == photo.id }) else { return }
          let newPhoto = self.createPhoto(photo)
          self.photos = self.photos.withReplaced(itemAt: index, newValue: newPhoto)
```

helecta commented on ImageFeed/Models/PhotosModel.swift:
```diff
@@ -0,0 +1,44 @@
+}
+
+struct LikeResult: Codable {
+    let isLike: Bool
```
LikeResult не такой:
```
struct LikeResult {
    let photo: PhotoResult
}
```

Peppa-Akk committed:
Sprint 12:
- Fix formatters;
- Fix optional createdAt;
- Fix Button Like;
- Add allert when logout.
